File: webui/frontend/src/lib/markdownSafe.ts
/**
 * Markdown-safe partial rendering for optional streaming (#220).
 *
 * Naive re-render of each SSE/WS chunk flashes unclosed `**`, `*`, `` ` ``,
 * fences, and links. This helper returns only the balanced prefix; the tail
 * stays buffered until closers arrive or the stream ends (`complete: true`).
 *
 * Invariant: `visible + held === source` (never drops text).
 */

export interface MarkdownSafePartial {
  visible: string
  held: string
}

export interface MarkdownSafeOptions {
  /** When true (generation ended), flush the held tail as-is. */
  complete?: boolean
}

type Kind =
  | 'strong-star'
  | 'em-star'
  | 'strong-under'
  | 'em-under'
  | 'code'
  | 'fence'
  | 'link'
  | 'linkdest'

interface Frame {
  kind: Kind
  start: number
  ticks?: number
  fenceChar?: string
  fenceLen?: number
}

function countRun(src: string, i: number, ch: string): number {
  let n = 0
  while (i + n < src.length && src[i + n] === ch) n++
  return n
}

function isLineStart(src: string, i: number): boolean {
  return i === 0 || src[i - 1] === '\n'
}

function matchFenceClose(
  src: string,
  i: number,
  fenceChar: string,
  fenceLen: number,
): number {
  let j = i
  let spaces = 0
  while (j < src.length && src[j] === ' ' && spaces < 3) {
    j++
    spaces++
  }
  if (j >= src.length || src[j] !== fenceChar) return -1
  const run = countRun(src, j, fenceChar)
  if (run < fenceLen) return -1
  j += run
  while (j < src.length && (src[j] === ' ' || src[j] === '\t')) j++
  if (j === src.length || src[j] === '\n') {
    return j < src.length ? j + 1 : j
  }
  return -1
}

function popKind(stack: Frame[], kind: Kind): boolean {
  for (let s = stack.length - 1; s >= 0; s--) {
    if (stack[s].kind === kind) {
      stack.splice(s, 1)
      return true
    }
  }
  return false
}

function oddTrailingBackslash(src: string): number {
  const n = src.length
  if (n === 0 || src[n - 1] !== '\\') return n
  let k = n - 1
  while (k >= 0 && src[k] === '\\') k--
  const count = n - 1 - k
  return count % 2 === 1 ? n - 1 : n
}

function firstHeldIndex(src: string): number {
  const n = src.length
  const stack: Frame[] = []
  let i = 0

  const top = (): Frame | undefined => stack[stack.length - 1]

  while (i < n) {
    const ch = src[i]
    const t = top()

    if (t?.kind === 'fence') {
      if (isLineStart(src, i)) {
        const close = matchFenceClose(src, i, t.fenceChar!, t.fenceLen!)
        if (close >= 0) {
          stack.pop()
          i = close
          continue
        }
      }
      i++
      continue
    }

    if (t?.kind === 'code') {
      if (ch === '`') {
        const run = countRun(src, i, '`')
        if (run === t.ticks) {
          stack.pop()
          i += run
          continue
        }
      }
      i++
      continue
    }

    if (t?.kind === 'linkdest') {
      if (ch === ')') {
        stack.pop()
        i++
        continue
      }
      i++
      continue
    }

    if (ch === '\\' && i + 1 < n) {
      i += 2
      continue
    }

    if (isLineStart(src, i)) {
      let j = i
      let spaces = 0
      while (j < n && src[j] === ' ' && spaces < 3) {
        j++
        spaces++
      }
      if (j < n && (src[j] === '`' || src[j] === '~')) {
        const fenceChar = src[j]
        const run = countRun(src, j, fenceChar)
        if (run >= 3) {
          let k = j + run
          while (k < n && src[k] !== '\n') k++
          stack.push({ kind: 'fence', start: i, fenceChar, fenceLen: run })
          i = k < n ? k + 1 : k
          continue
        }
        if (fenceChar === '~' && j + run === n) {
          stack.push({ kind: 'fence', start: i, fenceChar, fenceLen: 3 })
          break
        }
      }
    }

    if (ch === '`') {
      const run = countRun(src, i, '`')
      stack.push({ kind: 'code', start: i, ticks: run })
      i += run
      continue
    }

    if (ch === '!' && i + 1 < n && src[i + 1] === '[') {
      stack.push({ kind: 'link', start: i })
      i += 2
      continue
    }
    if (ch === '[') {
      stack.push({ kind: 'link', start: i })
      i++
      continue
    }
    if (ch === ']' && t?.kind === 'link') {
      if (i + 1 < n && src[i + 1] === '(') {
        t.kind = 'linkdest'
        i += 2
        continue
      }
      stack.pop()
      i++
      continue
    }

    if (ch === '*' || ch === '_') {
      const run = countRun(src, i, ch)
      if (
        isLineStart(src, i) &&
        ch === '*' &&
        run === 1 &&
        (i + 1 >= n || src[i + 1] === ' ' || src[i + 1] === '\t')
      ) {
        i++
        continue
      }
      let remaining = run
      let pos = i
      const strongKind: Kind = ch === '*' ? 'strong-star' : 'strong-under'
      const emKind: Kind = ch === '*' ? 'em-star' : 'em-under'
      const prev = i > 0 ? src[i - 1] : ''
      const canClose = prev !== '' && prev !== ' ' && prev !== '\t' && prev !== '\n'
      const next = i + run < n ? src[i + run] : ''
      const canOpen = next !== ' ' && next !== '\t' && next !== '\n'
      const openAtEos = next === ''

      while (remaining > 0) {
        if (remaining >= 2 && canClose && popKind(stack, strongKind)) {
          remaining -= 2
          pos += 2
          continue
        }
        if (remaining >= 1 && canClose && popKind(stack, emKind)) {
          remaining -= 1
          pos += 1
          continue
        }
        if (remaining >= 2 && (canOpen || openAtEos)) {
          stack.push({ kind: strongKind, start: pos })
          remaining -= 2
          pos += 2
          continue
        }
        if (remaining >= 1 && (canOpen || openAtEos)) {
          stack.push({ kind: emKind, start: pos })
          remaining -= 1
          pos += 1
          continue
        }
        remaining -= 1
        pos += 1
      }
      i += run
      continue
    }

    i++
  }

  let cut = n
  if (stack.length > 0) cut = Math.min(cut, stack[0].start)
  cut = Math.min(cut, oddTrailingBackslash(src))
  return cut
}

/** Split `source` into a renderable prefix and a held tail. */
export function markdownSafePartial(
  source: string,
  options?: MarkdownSafeOptions,
): MarkdownSafePartial {
  const text = String(source ?? '')
  if (!text) return { visible: '', held: '' }
  if (options?.complete) return { visible: text, held: '' }
  const cut = firstHeldIndex(text)
  if (cut >= text.length) return { visible: text, held: '' }
  if (cut <= 0) return { visible: '', held: text }
  return { visible: text.slice(0, cut), held: text.slice(cut) }
}

/**
 * Safe prefix of a (possibly incomplete) markdown stream.
 * Pass `{ complete: true }` at generation end to flush the remainder.
 */
export function renderMarkdownSafe(source: string, options?: MarkdownSafeOptions): string {
  return markdownSafePartial(source, options).visible
}

File: src/swarm/tui/markdown_safe.py
"""Markdown-safe partial rendering for optional streaming (#220).

Mirrors ``webui/frontend/src/lib/markdownSafe.ts``. Hold unclosed bold,
italic, inline code, fenced blocks, and links until the closer arrives or
the stream ends (``complete=True``). Never drops text.
"""

from __future__ import annotations

from typing import NamedTuple


class MarkdownSafePartial(NamedTuple):
    visible: str
    held: str


_STRONG_STAR = "strong-star"
_EM_STAR = "em-star"
_STRONG_UNDER = "strong-under"
_EM_UNDER = "em-under"
_CODE = "code"
_FENCE = "fence"
_LINK = "link"
_LINKDEST = "linkdest"


def _count_run(src: str, i: int, ch: str) -> int:
    n = 0
    length = len(src)
    while i + n < length and src[i + n] == ch:
        n += 1
    return n


def _is_line_start(src: str, i: int) -> bool:
    return i == 0 or src[i - 1] == "\n"


def _match_fence_close(src: str, i: int, fence_char: str, fence_len: int) -> int:
    j = i
    spaces = 0
    length = len(src)
    while j < length and src[j] == " " and spaces < 3:
        j += 1
        spaces += 1
    if j >= length or src[j] != fence_char:
        return -1
    run = _count_run(src, j, fence_char)
    if run < fence_len:
        return -1
    j += run
    while j < length and src[j] in " \t":
        j += 1
    if j == length or src[j] == "\n":
        return j + 1 if j < length else j
    return -1


def _pop_kind(stack: list[dict], kind: str) -> bool:
    for s in range(len(stack) - 1, -1, -1):
        if stack[s]["kind"] == kind:
            del stack[s]
            return True
    return False


def _odd_trailing_backslash(src: str) -> int:
    n = len(src)
    if n == 0 or src[-1] != "\\":
        return n
    k = n - 1
    while k >= 0 and src[k] == "\\":
        k -= 1
    count = n - 1 - k
    return n - 1 if count % 2 == 1 else n


def _first_held_index(src: str) -> int:
    n = len(src)
    stack: list[dict] = []
    i = 0

    def top() -> dict | None:
        return stack[-1] if stack else None

    while i < n:
        ch = src[i]
        t = top()

        if t is not None and t["kind"] == _FENCE:
            if _is_line_start(src, i):
                close = _match_fence_close(src, i, t["fence_char"], t["fence_len"])
                if close >= 0:
                    stack.pop()
                    i = close
                    continue
            i += 1
            continue

        if t is not None and t["kind"] == _CODE:
            if ch == "`":
                run = _count_run(src, i, "`")
                if run == t["ticks"]:
                    stack.pop()
                    i += run
                    continue
            i += 1
            continue

        if t is not None and t["kind"] == _LINKDEST:
            if ch == ")":
                stack.pop()
                i += 1
                continue
            i += 1
            continue

        if ch == "\\" and i + 1 < n:
            i += 2
            continue

        if _is_line_start(src, i):
            j = i
            spaces = 0
            while j < n and src[j] == " " and spaces < 3:
                j += 1
                spaces += 1
            if j < n and src[j] in "`~":
                fence_char = src[j]
                run = _count_run(src, j, fence_char)
                if run >= 3:
                    k = j + run
                    while k < n and src[k] != "\n":
                        k += 1
                    stack.append(
                        {
                            "kind": _FENCE,
                            "start": i,
                            "fence_char": fence_char,
                            "fence_len": run,
                        }
                    )
                    i = k + 1 if k < n else k
                    continue
                if fence_char == "~" and j + run == n:
                    stack.append(
                        {
                            "kind": _FENCE,
                            "start": i,
                            "fence_char": fence_char,
                            "fence_len": 3,
                        }
                    )
                    break

        if ch == "`":
            run = _count_run(src, i, "`")
            stack.append({"kind": _CODE, "start": i, "ticks": run})
            i += run
            continue

        if ch == "!" and i + 1 < n and src[i + 1] == "[":
            stack.append({"kind": _LINK, "start": i})
            i += 2
            continue
        if ch == "[":
            stack.append({"kind": _LINK, "start": i})
            i += 1
            continue
        if ch == "]" and t is not None and t["kind"] == _LINK:
            if i + 1 < n and src[i + 1] == "(":
                t["kind"] = _LINKDEST
                i += 2
                continue
            stack.pop()
            i += 1
            continue

        if ch in "*_":
            run = _count_run(src, i, ch)
            if (
                _is_line_start(src, i)
                and ch == "*"
                and run == 1
                and (i + 1 >= n or src[i + 1] in " \t")
            ):
                i += 1
                continue
            remaining = run
            pos = i
            strong_kind = _STRONG_STAR if ch == "*" else _STRONG_UNDER
            em_kind = _EM_STAR if ch == "*" else _EM_UNDER
            prev = src[i - 1] if i > 0 else ""
            can_close = prev not in ("", " ", "\t", "\n")
            nxt = src[i + run] if i + run < n else ""
            can_open = nxt not in (" ", "\t", "\n")
            open_at_eos = nxt == ""

            while remaining > 0:
                if remaining >= 2 and can_close and _pop_kind(stack, strong_kind):
                    remaining -= 2
                    pos += 2
                    continue
                if remaining >= 1 and can_close and _pop_kind(stack, em_kind):
                    remaining -= 1
                    pos += 1
                    continue
                if remaining >= 2 and (can_open or open_at_eos):
                    stack.append({"kind": strong_kind, "start": pos})
                    remaining -= 2
                    pos += 2
                    continue
                if remaining >= 1 and (can_open or open_at_eos):
                    stack.append({"kind": em_kind, "start": pos})
                    remaining -= 1
                    pos += 1
                    continue
                remaining -= 1
                pos += 1
            i += run
            continue

        i += 1

    cut = n
    if stack:
        cut = min(cut, stack[0]["start"])
    cut = min(cut, _odd_trailing_backslash(src))
    return cut


def markdown_safe_partial(source: str, *, complete: bool = False) -> MarkdownSafePartial:
    text = "" if source is None else str(source)
    if not text:
        return MarkdownSafePartial("", "")
    if complete:
        return MarkdownSafePartial(text, "")
    cut = _first_held_index(text)
    if cut >= len(text):
        return MarkdownSafePartial(text, "")
    if cut <= 0:
        return MarkdownSafePartial("", text)
    return MarkdownSafePartial(text[:cut], text[cut:])


def render_markdown_safe(source: str, *, complete: bool = False) -> str:
    """Safe prefix of a (possibly incomplete) markdown stream."""
    return markdown_safe_partial(source, complete=complete).visible


__all__ = [
    "MarkdownSafePartial",
    "markdown_safe_partial",
    "render_markdown_safe",
]
